### ISSUE TITLE
Implementing Bias and tanh/softmax activation function

### ISSUE THREAD

**Issue by sunilson** [makeyourownneuralnetwork/makeyourownneuralnetwork#47]
I am trying to extend the example from your book with those things, but I can't get them to work. Can you maybe show how that would be done in python?

Thanks!

**Comment by Gerrit02** [comment]
Hello @sunilson 
maybe this can help: 
http://makeyourownneuralnetwork.blogspot.be/2016/06/bias-nodes-in-neural-networks.html

**Comment by Gerrit02** [comment]
Maybe if you have some time left you can try what I explained in post (see below) and tell me what you have as a result?
https://github.com/makeyourownneuralnetwork/makeyourownneuralnetwork/issues/46

**Comment by sunilson** [comment]
I already read that but I couldn't really get it to work. I then tried the approach of Shiffman that he does here: https://github.com/CodingTrain/Toy-Neural-Network-JS/blob/master/lib/nn.js

So he just has some extra matrices which contain the bias values and adds the gradient to them of the respective weight. (See line 117 for example). But I am not sure if this is the correct approach?

**Comment by Gerrit02** [comment]
Hello @sunilson 
Sorry but I think I can not help you with that. I'm also beginning learning about neural networks and wanted also extracted the code with bias and saving the network etc... Rashid responded already he has not the time to provide this or extend his code with it. The only thing we have is the links he provided about biasses and saving the network.
What I did is I ordered me a book via amazon.com (Neural Network Programming with Python) and when I see the content of that book, it will go more in dept. So far I didn't receive it yet.

Did you find already some time to look or try what I explained in 
https://github.com/makeyourownneuralnetwork/makeyourownneuralnetwork/issues/46

**Comment by makeyourownneuralnetwork** [comment]
Hi @sunilson 

Sorry I've not had a chance to reply to you sooner.

If you want to use a different activation function then you need to work through the calculus to work out how the error depends on the weights (dE/dw). In the book we use the sigmoid function and the calculus algebra is easy.

For tanh() you can find examples of the calculus in many textbooks, here are some examples: 
 - [https://theclevermachine.wordpress.com/tag/backpropagation-algorithm/](https://theclevermachine.wordpress.com/tag/backpropagation-algorithm/)
 - [https://www.inf.ed.ac.uk/teaching/courses/mlp/2017-18/mlp03-mlp.pdf](https://www.inf.ed.ac.uk/teaching/courses/mlp/2017-18/mlp03-mlp.pdf)

Ultimately this will mean your updating code new_weight = old_weight + something will be different for different activation functions.

Regarding bias - the easiest way to implement them is to add an extra node in the input and hidden layers which always has a value of 1.0. This way you don't need to change anything in terms of the maths for back propagation. This way, the weights associated with the bias node are trained and you get the benefit of a non-zero bias.

**Comment by makeyourownneuralnetwork** [comment]
I will try hard to find some time to implement the code to illustrate the use of something like tang() and also bias nodes too .. but sadly right now I am working on my next books.

**Comment by sunilson** [comment]
No worries, and thanks for your answer!
